### ISSUE TITLE
Agent constantly disconnects from server after 1 minute

### ISSUE THREAD

**Issue by y8** [rancher/rancher#5086]
**Rancher Version:** 1.0.2

**Docker Version:** 1.10.3, build 20f81dd

**OS and where are the hosts located? (cloud, bare metal, etc):** bare metal
`#38~14.04.1-Ubuntu SMP Fri Nov 6 18:17:28 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux`

**Setup Details: (single node rancher vs. HA rancher, internal DB vs. external DB)** external db (percona)

**Environment Type: (Cattle/Kubernetes/Swarm/Mesos)** Cattle

**Steps to Reproduce:**  (see notes)

**Notes:**

One of the our environments is deployed on the same host as `rancher/server`. Agent constantly disconnects after 1 minute 15 seconds.

rancher/agent:v1.0.1 log:

```
I0613 16:34:11.414530   15869 cadvisor.go:106] Starting cAdvisor version: 0.22.0- on port 9344
2016-06-13 14:35:31,092 ERROR agent [140592204339312] [event.py:201] Max of [10] dropped [ping] requests exceeded 
```

Here a full log: https://gist.github.com/y8/ce681cf002c61c4f564d0b8d59a08d31#file-rancher-agent-log

rancher/server:v1.0.2 log
```
2016-06-13 14:35:16,933 ERROR [:] [] [] [] [ecutorService-2] [i.c.p.a.s.ping.impl.PingMonitorImpl ] Failed to get ping from agent [5] count [3] 
2016-06-13 14:35:21,933 ERROR [:] [] [] [] [cutorService-10] [i.c.p.a.s.ping.impl.PingMonitorImpl ] Failed to get ping from agent [5] count [4] 
2016-06-13 14:35:26,934 ERROR [:] [] [] [] [ecutorService-6] [i.c.p.a.s.ping.impl.PingMonitorImpl ] Failed to get ping from agent [5] count [5] 
2016-06-13 14:35:31,934 ERROR [:] [] [] [] [ecutorService-9] [i.c.p.a.s.ping.impl.PingMonitorImpl ] Failed to get ping from agent [5] count [6] 
2016-06-13 14:35:31,935 ERROR [:] [] [] [] [ecutorService-9] [i.c.p.a.s.ping.impl.PingMonitorImpl ] Scheduling reconnect for [5] 
2016-06-13 14:35:36,935 ERROR [:] [] [] [] [ecutorService-3] [i.c.p.a.s.ping.impl.PingMonitorImpl ] Failed to get ping from agent [5] count [7] 
2016-06-13 14:35:41,935 ERROR [:] [] [] [] [ecutorService-5] [i.c.p.a.s.ping.impl.PingMonitorImpl ] Failed to get ping from agent [5] count [8] 
time="2016-06-13T14:35:44Z" level=info msg="Registering backend for host [d0ca9a5c-510f-44f0-aef4-b65f737004a5]" 
2016-06-13 14:35:46,936 ERROR [:] [] [] [] [ecutorService-5] [i.c.p.a.s.ping.impl.PingMonitorImpl ] Failed to get ping from agent [5] count [9] 
```

1. In the `Infrastructure` tab agent report correct public IP for the host. 
2. For 1 minute exactly agent work as expected and can schedule containers. Console / logs for containers are working as expected, but connection drops when agent disconnects. Same happens with stats
3. We are using firewall to allow only 500/4500 upd between rancher nodes, but I've tried to disable ufw completely for a while, same issue occurs.
4. This affect only one environment, another environment is deployed to different hosts and behaves as expected
5. We are using nginx to terminate SSL connections to rancher, we are using the nginx config from the docs, but just in case, here it is: https://gist.github.com/y8/ce681cf002c61c4f564d0b8d59a08d31#file-nginx-conf. Nginix is running in container and linked to `rancher-server`. Changing the `proxy_read_timeout` value doesn't affect the 1m15s reconnection interval.
6. This agent show `0 bytes` in the `Storage` for the host, agents in other environment are showing the correct values.

**Comment by Rucknar** [comment]
What do you have your cattle url set to? The strings hosts connect to when you add the agent?

**Comment by y8** [comment]
> What do you have your cattle url set to? The strings hosts connect to when you add the agent?
You can see from the logs: `https://<rancher-server-host>.<project.tld>:444/`. As I said:
1. Agent behaves correctly exactly for 1 minute
2. Second environment works without a single glitch

**Comment by cloudnautique** [comment]
why is it using port 444? The nginx server config is listening on 443. Rancher has an interesting trick where it can do HTTP/HTTPS on the same port, and I'm guessing (purely a guess) that your not really talking over SSL to rancher server. I would set your registration url to https://<rancher-server-host>.<project.tld> in the admin settings and strip port 444.

**Comment by y8** [comment]
Because ports 80 and 443 are used by LoadBalancer on this host. 

Port 444 is exposed by `nginx` container, that linked to `rancher-server` to terminate SSL. Accessing this registration url in Chrome passes all HTTPS checks. Registration url for agent matches the configuration in the admin settings.

Besides, if that was a cause, we would have similar issues with other nodes connected to this rancher server. We have 5 bare metal servers and only one (with "rancher-server" on it) have this "blinking" agent issue.

As you can see from agent logs it complains about `ECONNREFUSED` to `cAdvisor` a lot. Maybe this is a cause? @vincent99 told me that cAdvisor stuff is a part of the ping check. 

Btw, we haven't had this issue before 1.0.x upgrade on the exactly same configuration, including 444 port for registration.

**Comment by cloudnautique** [comment]
Thanks for the additional info, the 443/444 thing was throwing me for a loop without knowing your environments full context. The HTTP/HTTPS Cattle trick is real subtle and can cause issues. 

Yes, cAdvisor and docker inspects are returned as part of the ping responses. If cAdvisor is unavailable, we should handle that pretty well... if docker is slow we start missing pings like crazy. 

The `ECONNREFUSED` errors are common when an agent starts up. cAdvisor restarts and python-agent tries to grab stats before it is fully running We need to catch that better somewhere. 

This is the only host in 5 that has issues? We saw this behavior with a default timeout setting in nginx not being correct, but that would mean all of them would do this. We also added keepalive code in the agents to not require it. 

Can you try registering the host with an env var `-e CATTLE_SCRIPT_DEBUG=true` see if that gives us any more info.

**Comment by y8** [comment]
I've restarted agent in debug mode, but there exactly nothing new about the what exactly caused the ping failure. 

```
I0617 18:35:35.307045    7691 factory.go:233] Registering Docker factory
I0617 18:35:35.308365    7691 factory.go:97] Registering Raw factory
I0617 18:35:38.341912    7691 manager.go:1039] Started watching for new ooms in manager
I0617 18:35:38.341950    7691 oomparser.go:198] OOM parser using kernel log file: "/var/log/kern.log"
I0617 18:35:38.343824    7691 manager.go:292] Starting recovery of all containers
I0617 18:35:38.449766    7691 manager.go:297] Recovery completed
I0617 18:35:38.470535    7691 cadvisor.go:106] Starting cAdvisor version: 0.22.0- on port 9344
2016-06-17 16:36:57,122 ERROR agent [140031819606128] [event.py:201] Max of [10] dropped [ping] requests exceeded 
[run.sh:317] sleep 5
```

And then it just a re-connection output with a lot of debugging info. Here a full log: https://gist.github.com/y8/ee8aaf11b49dde5e02d444da28a10a7b

**Comment by y8** [comment]
Also this happens time to time: https://gist.github.com/y8/819f099e3f468edba3f83cf196abd034

**Comment by y8** [comment]
By the way, dunno if this is related, but failing environment have strange `uuid` property:

```json
  "uuid": "adminProject"
```

Second environment got the proper one ("zzzzzzzz-xxxx-xxxx-xxxx-yyyyyyyyyyyy"). Could it be something like upgrade went wrong?

**Comment by dkiser** [comment]
Minus the panic, seeing exact same issues on following

**Rancher version:** 1.0.2

**docker version**
```
Client:
 Version:      1.10.3
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   20f81dd
 Built:        Thu Mar 10 15:39:25 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.10.3
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   20f81dd
 Built:        Thu Mar 10 15:39:25 2016
 OS/Arch:      linux/amd64
```

**os version**
```
CentOS Linux release 7.2.1511 (Core)
````

**Comment by y8** [comment]
@cloudnautique @alena1108 could you please add the milestone and labels to this ticket?

**Comment by y8** [comment]
@deniseschannon or @sangeethah maybe you can label and milestone this one?

**Comment by vincent99** [comment]
@y8 claims (in IRC) that this is fixed in 1.1.
